[pt][quant] Unify numerics between fakequant and quant/dequant

dskhudia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37188 [pt][quant] Unify numerics between fakequant and quant/dequant**

Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

dskhudia committed:
[pt][quant] Unify numerics between fakequant and quant/dequant

Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Unify numerics between fakequant and quant/dequant"


Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -209,7 +205,10 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
     // using std::round because it does rounding away from zero in halfway
     // cases.
-    dst[i] = nearbyint(clipped);
+    transformed = zero_point + nearbyint(transformed);
```
Based on the comment above, why cant we force the rounding mode to be FE_TONEAREST?

dskhudia replied:
> This is the default mode. It should be used unless there is a specific need for one of the others.
> You should avoid changing the rounding mode if possible. It can be an expensive operation; also, some hardware requires you to compile your program differently for it to work. The resulting code may run slower. 
https://www.gnu.org/software/libc/manual/html_node/Rounding.html

raghuramank100 replied:
Can we atleast check what the rounding mode is print a warning?

dskhudia replied:
Is there a central init routine where we can add such a warning? cc: @jerryzh168

raghuramank100 replied:
cc @dzhulgakov

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -1841,7 +1841,7 @@ void fake_quantize_tensor_kernel(
                 std::fmax(
                     static_cast<int64_t>(
-                        std::nearbyint(self * inv_scale + z_point)),
+                        z_point + std::nearbyint(self * inv_scale)),
```
Does the cast need to be to int64 _t here?

raghuramank100 commented on aten/src/ATen/test/quantized_test.cpp:
```diff
@@ -68,23 +68,21 @@ TEST(TestQTensor, QuantDequantAPIs) {
   for (size_t idx = 0; idx < x_values.size(); ++idx) {
     ASSERT_EQ(qx_expect[idx], qx_data[idx].val_)
-      << "Tie breaking during rounding element " << idx << " failed!";
+        << "Tie breaking during rounding element " << idx << " failed!";
```
In addition we need to ensure that the tests in test_fake_quant.py continue to work, likely the numerics would have changed.
In addition, all the backward compability tests for serialization will require regeneration of the serialized models and outputs.

dskhudia replied:
Are these tests already running in CI?

dskhudia replied:
All the tests in https://github.com/pytorch/pytorch/blob/master/test/quantization/test_backward_compatibility.py and https://github.com/pytorch/pytorch/blob/master/test/quantization/test_fake_quant.py are passing in my local run.

dskhudia committed:
Update on "[pt][quant] Unify numerics between fakequant and quant/dequant"


Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Unify numerics between fakequant and quant/dequant"


Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Unify numerics between fakequant and quant/dequant"


Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Unify numerics between fakequant and quant/dequant"


Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Unify numerics between fakequant and quant/dequant"


Add zero point after rounding in both fakequant and quant.

Differential Revision: [D21193552](https://our.internmc.facebook.com/intern/diff/D21193552/)

[ghstack-poisoned]
